### PR TITLE
Document support under Autograph for JAX array assignment

### PR DESCRIPTION
**Pull request by rauletorresc**
**Context:** The new feature must be documented as an entry to the @qjit docstring and the Autograph guide

[sc-60314]

### CODE REVIEW

**Review by erick-xanadu — APPROVED** [review]

File: doc/dev/autograph.rst [hunk 1]
```diff
@@ -896,3 +896,33 @@ the context will be converted anyways:
...
+... def f(x):
+...     first_dim = x.shape[0]
+...     result = jnp.empty((first_dim,), dtype=x.dtype)
+...
+...     for i in range(first_dim):
+...         result = result.at[i].set(x[i]* 2)
+...
+...     return result 
+    
+However, Catalyst has enabled additional support for the single index array assignment syntax we all know:
```
Comment (josh146) [inline doc/dev/autograph.rst:916]:
```suggestion
However, if updating a single index of the array, Autograph supports conversion of standard Python array assignment syntax:
```

File: doc/dev/autograph.rst [hunk 2]
```diff
@@ -896,3 +896,33 @@ the context will be converted anyways:
 ...     with disable_autograph:
 ...         x += f()
 ...     return x
+
+
+Dealing with JAX array assignments inside Autograph
+---------------------------------------------------
```
Comment (josh146) [inline doc/dev/autograph.rst:902]:
```suggestion
In-place array assignments
--------------------------
```

File: doc/dev/autograph.rst [hunk 3]
```diff
@@ -896,3 +896,33 @@ the context will be converted anyways:
...
+
+By default, updating a JAX array item is only possible by using the 'at' and 'set' methods:
```
Comment (josh146) [inline doc/dev/autograph.rst:904]:
```suggestion
To update array values when using JAX, the
`JAX syntax for array assignment <https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#array-updates-x-at-idx-set-y>`__
which uses the array ``at`` and ``set`` methods must be used:
```

File: frontend/catalyst/jit.py [hunk 4]
```diff
@@ -581,6 +581,41 @@ def circuit(x: int):
         be unrolled during tracing, "copy-pasting" the body 5 times into the program rather than
         appearing as is.
 
+
+    .. details::
+        :title: Dealing with JAX array assignments inside Autograph
```
Comment (josh146) [inline frontend/catalyst/jit.py:586]:
```suggestion
        :title: In-place array assignment
```

File: frontend/catalyst/jit.py [hunk 5]
```diff
@@ -581,6 +581,41 @@ def circuit(x: int):
...
+        .. code-block:: python
+            @qjit(autograph=True)
+            def f(x):
+            first_dim = x.shape[0]
+            result = jnp.empty((first_dim,), dtype=x.dtype)
+
+            for i in range(first_dim):
+                result = result.at[i].set(x[i]* 2)
+
+            return result
```
Comment (josh146) [inline frontend/catalyst/jit.py:599]:
```suggestion
        .. code-block:: python

            @qjit(autograph=True)
            def f(x):
            first_dim = x.shape[0]
            result = jnp.empty((first_dim,), dtype=x.dtype)

            for i in range(first_dim):
                result = result.at[i].set(x[i]* 2)

            return result
```

File: frontend/catalyst/jit.py [hunk 6]
```diff
@@ -581,6 +581,41 @@ def circuit(x: int):
...
+        .. code-block:: python
+            @qjit(autograph=True)
+            def f(x):
+            first_dim = x.shape[0]
+            result = jnp.empty((first_dim,), dtype=x.dtype)
+
+            for i in range(first_dim):
+                result[i] = x[i] * 2
+
+            return result
```
Comment (josh146) [inline frontend/catalyst/jit.py:613]:
```suggestion
        .. code-block:: python

            @qjit(autograph=True)
            def f(x):
            first_dim = x.shape[0]
            result = jnp.empty((first_dim,), dtype=x.dtype)

            for i in range(first_dim):
                result[i] = x[i] * 2

            return result
```

File: frontend/catalyst/jit.py [hunk 7]
```diff
@@ -581,6 +581,41 @@ def circuit(x: int):
...
+            first_dim = x.shape[0]
+            result = jnp.empty((first_dim,), dtype=x.dtype)
+
+            for i in range(first_dim):
+                result = result.at[i].set(x[i]* 2)
+
+            return result
+
+        However, Catalyst has enabled additional support for the single index array assignment
+        syntax we all know:
```
Comment (josh146) [inline frontend/catalyst/jit.py:602]:
```suggestion
        However, if updating a single index of the array, Autograph supports conversion of standard Python array assignment syntax:
```

File: frontend/catalyst/jit.py [hunk 8]
```diff
@@ -581,6 +581,41 @@ def circuit(x: int):
         be unrolled during tracing, "copy-pasting" the body 5 times into the program rather than
         appearing as is.
 
+
+    .. details::
+        :title: Dealing with JAX array assignments inside Autograph
+
+        By default, updating a JAX array item is only possible using the 'at' and 'set' methods:
```
Comment (josh146) [inline frontend/catalyst/jit.py:588]:
```suggestion
        To update array values when using JAX, the
`JAX syntax for array assignment <https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#array-updates-x-at-idx-set-y>`__
which uses the array ``at`` and ``set`` methods must be used:
```

**Review by josh146 — APPROVED** [review]
thanks @rauletorresc, left some suggestions!

File: doc/dev/autograph.rst (hunk 1 above)
Comment (rauletorresc) [reply doc/dev/autograph.rst]: Done

File: doc/dev/autograph.rst (hunk 2 above)
Comment (rauletorresc) [reply doc/dev/autograph.rst]: Done

File: doc/dev/autograph.rst (hunk 3 above)
Comment (rauletorresc) [reply doc/dev/autograph.rst]: Done

File: frontend/catalyst/jit.py (hunk 4 above)
Comment (rauletorresc) [reply frontend/catalyst/jit.py]: Done

File: frontend/catalyst/jit.py (hunk 5 above)
Comment (rauletorresc) [reply frontend/catalyst/jit.py]: Done

File: frontend/catalyst/jit.py (hunk 8 above)
Comment (rauletorresc) [reply frontend/catalyst/jit.py]: Done

**Pull request merged**